### PR TITLE
correctly display '&' and '&nbsp;' in product descriptions

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Addresses #6285 

<s>This is a pretty hacky fix and I don't think we'd want to merge it as is.... just putting it out there as a draft</s>. This might be our simplest option for getting 80% of the way there to not having users see "weird" stuff in their product descriptions. It's limited to the first and last examples given in the issue (doesn't address strings like `<!-- Start Fragment -->`). 

It's definitely a bandaid solution and ideally we'd take a look at the behavior of that WYSIWYG editor to see if that's a cause of confusion. But depending on what other people think, this could be a decent stopgap. 

#### What should we test?

A product with `&amp;` and a product with `&nbsp;` in the description should not show those characters on the shop front but instead translate them to `&` and ` ` (blank space) respectively.

#### Release notes
Bug fix: product descriptions no longer show escaped HTML fragments. 

Changelog Category: User facing changes

### CODE REVIEW

**Comment by RachL** [comment]
@andrewpbrett hello :) should we keep this PR open?

**Comment by andrewpbrett** [comment]
@RachL Yes, I had another look and the Rails 4.2 upgrade did not in fact get us anything for free... but I thought through the issue some more and I think we should move this to Code Review and see what other people think, if it's an okay solution for now that will take care of the worse offenses.

**Review by Matt-Yorkley — APPROVED** [review]

File: app/serializers/api/product_serializer.rb [hunk 1]
```diff
@@ -16,16 +16,30 @@ class Api::ProductSerializer < ActiveModel::Serializer
...
   end
 
   # return a sanitized html description
   def description_html
     d = sanitize(object.description, tags: ["p", "b", "strong", "em", "i", "a", "u"],
                                      attributes: ["href", "target"])
-    d.to_s.html_safe
+    d = d.to_s.html_safe
+    ALLOWED_CHARACTERS.each do |character, sub|
+      d = d.gsub(character, sub)
```
Comment (Matt-Yorkley) [inline app/serializers/api/product_serializer.rb:40]: There's quite a bit of logic in this Serializer now for sanitizing HTML. It feels like it shouldn't be this class's responsibility. Maybe we could pull these bits into a small service object?

Comment (andrewpbrett) [reply app/serializers/api/product_serializer.rb]: 03ede57b32f8066ee7bf6109145a47172d180eeb

**Comment by Matt-Yorkley** [comment]
I made a little PR here: https://github.com/andrewpbrett/openfoodnetwork/pull/22

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by sauloperez** [comment]
It'd be great if we could add some unit tests to `ContentSanitizer` while this moves to Test Ready :pray:

**Review by sauloperez — APPROVED** [review]

File: spec/services/content_sanitizer.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    it "replaces ampersands" do
+      expect(service.sanitize_content("pb &amp; j")).to eq("pb & j")
+    end
+
+    it "replaces double escaped ampersands" do
+      expect(service.sanitize_content("pb &amp;amp; j")).to eq("pb & j")
+    end
+  end
+end
```
Comment (sauloperez) [inline spec/services/content_sanitizer.rb:49]: :heart:

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

There's something not quite right with this one. Right after staging, it seems no OC is loading produts - I tried all hubs on staging-UK (endless spinner), all of them showed the endless spinner, after choosing an open OC:

image

This Bugsnag is related:
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/5fff2ddbeef13a0019b77cd6?event_id=5fff2ddb00659d822fd60000&i=sk&m=nw

The only OC which loaded correctly was this one - `https://staging.openfoodnetwork.org.uk/la-jimena/shop` - so this can help perhaps to provide an answer on what needs to be improved on the code.

Restaging another PR after this one brought all the other OCs back to life.

I think we need to investigate - how come this did not break any spec related to displaying the products from an open OC in the  shopfront.

Moving back to In Dev for now.

**Comment by andrewpbrett** [comment]
Thanks @filipefurtad0, great catch. I added a commit that should address it (and another one for something else I noticed), so if the build is green, feel free to re-test. I don't think it needs an additional review.

**Comment by filipefurtad0** [comment]
Thanks for addressing this so quickly @andrewpbrett ,

It's working now - I had a peek at your tests and pasted some to check whether it had the expected effect:

image

Verified it to work without clicking on the product:
image

Good to go!

**Pull request merged**